### PR TITLE
[MD] Refactor data source server side error handling

### PR DESCRIPTION
**Pull request by zhongnansu**
Signed-off-by: Su <szhongna@amazon.com>

### Description
Refactor data source server side error handling
- Rename `DataSourceConfigError` to `DataSourceError`
- Add helper `createDataSourceError()` to create data source error from:
  - Opensearch client error
  - legacy client error
  - saved object client error
  - other runtime errors
- Replace  usage of `new DataSourceError(e)` with `createDataSourceError(e) `
- can't directly import `createDataSourceError()` to `data` plugin, have to import from data source plugin setup interface, otherwise, some other UTs in other modules fail
- refactor to move data source error response logic from `router.ts` to `resolve_index.ts`.  I check all our codebase, only the route handler defined  resolve_index is missing error handling. All other `internal/xxx/xxx` API has err handling. E.g `internal/search/`, `/internal/_msearch`, `/internal/index-pattern-manage/preview-script`
    ```
    catch (err) {
            return res.customError({
              statusCode: err.statusCode,
              body: {
                message: err.message,
                attributes: {
                  error: err.body.error,
                },
              },
            });
    ```

#### Error statusCode casting logic 
  - If it's response error, we **maintain the error statusCode**, except 401
  - for any 401 error, we cast to 400, because of #2591
  - if it's client non-response error(without statuscode), cast to 400,  while keeping the original error message, such as timeout error, connection error, no living connection error, etc. Because by design, datasource could be created without validation of connectivity. Besides, ds could be updated, we won't be aware of the data source side status, also couldn't keep track of ds domain status~~

 
### Issues Resolved
https://github.com/opensearch-project/OpenSearch-Dashboards/issues/2591
 
### Check List
- [x] All tests pass
  - [ ] `yarn test:jest`
  - [ ] `yarn test:jest_integration`
  - [ ] `yarn test:ftr`
- [ ] New functionality includes testing.
- [ ] New functionality has been documented.
- [x] Update [CHANGELOG.md](./../CHANGELOG.md)
- [x] Commits are signed per the DCO using --signoff

### CODE REVIEW

File: src/plugins/data/server/search/opensearch_search/opensearch_search_strategy.ts [hunk 1]
```diff
@@ -29,10 +29,13 @@
  */
 
 import { first } from 'rxjs/operators';
-import { SharedGlobalConfig, Logger } from 'opensearch-dashboards/server';
 import { SearchResponse } from 'elasticsearch';
 import { Observable } from 'rxjs';
 import { ApiResponse } from '@opensearch-project/opensearch';
+// eslint-disable-next-line @osd/eslint/no-restricted-paths
```
Comment (ashwin-pc) [inline src/plugins/data/server/search/opensearch_search/opensearch_search_strategy.ts:35]: Why are we using a restricted path here?

**Review by ashwin-pc — COMMENTED** [review]
> We cast all search error against data source to 400 bad request, while maintaining the original error message from OS.

This worries me a lil because it is the root of a similar issue we face in other places in Dashboards where we incorrectly assign the wrong error code for a response. 4xx errors are auth errors. 

1. Is there really no way to assign the right error response code for an error?
2. If not, why did you choose to use a 4xx error instead of a generic 5xx error?

File: src/plugins/data/server/search/opensearch_search/opensearch_search_strategy.ts [hunk 2]
```diff
@@ -86,8 +87,11 @@ export const opensearchSearchStrategyProvider = (
           rawResponse,
           ...getTotalLoaded(rawResponse._shards),
         };
-      } catch (e) {
+      } catch (e: any) {
```
Comment (kavilla) [inline src/plugins/data/server/search/opensearch_search/opensearch_search_strategy.ts:90]: can this be more typed?

File: src/plugins/data_source/server/lib/error.test.ts [hunk 3]
```diff
@@ -4,35 +4,35 @@
...
+    expect(new DataSourceError('test prefix: ', error)).toMatchObject({
       statusCode: 500,
       message: 'test prefix: test reason message',
     });
   });
 
   it('create from non savedObject error should always be a 400 error', () => {
     const error = new Error('test reason message');
-    expect(new DataSourceConfigError('test prefix: ', error)).toMatchObject({
+    expect(new DataSourceError('test prefix: ', error)).toMatchObject({
```
Comment (kavilla) [inline src/plugins/data_source/server/lib/error.test.ts:35]: shouldn't this fail based on the new constructor

Comment (zhongnansu) [reply src/plugins/data_source/server/lib/error.test.ts]: yeah, just updated the existing tests.

**Comment by zhongnansu** [comment]
> > We cast all search error against data source to 400 bad request, while maintaining the original error message from OS.
> 
> This worries me a lil because it is the root of a similar issue we face in other places in Dashboards where we incorrectly assign the wrong error code for a response. 4xx errors are auth errors.
> 
> 1. Is there really no way to assign the right error response code for an error?
> 2. If not, why did you choose to use a 4xx error instead of a generic 5xx error?

@ashwin-pc , thanks for the comments, really helpful. I re-worked on this PR. basically doing a refactor of existing error handling logic in data source feature. Now it will keep original status code. But for 401, we choose to cast to 400, due to the limitation of security dashboards plugin. For 5xx error, we cast to 400. Because it doesn't make sense to use 5xx, e.g user's data source domain is down, while trying to connect to it, we get 5xx. But there's nothing wrong with local OSD.

File: src/plugins/data/server/search/opensearch_search/opensearch_search_strategy.ts (hunk 2 above)
Comment (zhongnansu) [reply src/plugins/data/server/search/opensearch_search/opensearch_search_strategy.ts]: don't think so, but let me switch back to use `unknown`, better than `any` at least

**Comment by ashwin-pc** [comment]
@zhongnansu Thanks for that explanation! That logic makes sense to me.

File: src/core/server/http/router/router.ts [hunk 4]
```diff
@@ -312,8 +312,8 @@ export class Router implements IRouter {
   }
 }
 
-const isDataSourceConfigError = (error: any) => {
-  return error.constructor.name === 'DataSourceConfigError' && error.statusCode === 400;
+const isDataSourceError = (error: any) => {
+  return error.constructor.name === 'DataSourceError' && error.statusCode === 400;
```
Comment (zengyan-amazon) [inline src/core/server/http/router/router.ts:316]: can we use `instanceof`instead of `error.constructor.name` ?

**Review by zengyan-amazon — APPROVED** [review]

File: src/core/server/http/router/router.ts (hunk 4 above)
Comment (zhongnansu) [reply src/core/server/http/router/router.ts]: importing `DataSourceError` from plugin module to here(core module) doesn't sound like a good practice. But anyway, I did some refactor to move data source error response logic from `router.ts` to `resolve_index.ts`(/internal/index-pattern-manage/resolve-index).  

I checked the codebase, only the route handler defined  resolve_index is missing error handling. All other `internal/xxx/xxx` API has err handling. E.g `internal/search/`, `/internal/_msearch`, `/internal/index-pattern-manage/preview-script`

    ```
    catch (err) {
            return res.customError({
              statusCode: err.statusCode,
              body: {
                message: err.message,
                attributes: {
                  error: err.body.error,
                },
              },
            });
    ```   

@zengyan-amazon could you review again

File: src/plugins/index_pattern_management/server/routes/resolve_index.ts [hunk 5]
```diff
@@ -63,13 +63,25 @@ export function registerResolveIndexRoute(router: IRouter): void {
...
+        const result = await caller('transport.request', {
+          method: 'GET',
+          path: `/_resolve/index/${encodeURIComponent(req.params.query)}${
+            queryString ? '?' + new URLSearchParams(queryString).toString() : ''
+          }`,
+        });
+        return res.ok({ body: result });
+      } catch (err) {
+        return res.customError({
+          statusCode: err.statusCode || 500,
```
Comment (kristenTian) [inline src/plugins/index_pattern_management/server/routes/resolve_index.ts:76]: I thought the redirect happens during this API call, but we are not handling the error using the `createDataSourceError` here?

Comment (zhongnansu) [reply src/plugins/index_pattern_management/server/routes/resolve_index.ts:76]: 401 is causing refresh by security plugin. Prior to here, we use `createDataSourceError()` to cast the 401 to 400, here it receives that 400, and will never get any 401 and won't trigger any refresh

Comment (kristenTian) [reply src/plugins/index_pattern_management/server/routes/resolve_index.ts:76]: I missed it, it is thrown during constructing the client

**Review by kristenTian — APPROVED** [review]

**Review by zengyan-amazon — APPROVED** [review]

**Pull request merged**